### DISCUSSION TITLE
Any example for Custom Controls?

### DISCUSSION THREAD

**Discussion by zehawki** [react-page/react-page#1162] [Q&A]
@macrozone any example to see, to understand how to get Custom Controls (https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-controls) to work.

**Comment by macrozone** [comment]
> @macrozone any example to see, to understand how to get Custom Controls (https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-controls) to work.

the built in image plugin uses custom controls (historically): https://github.com/react-page/react-page/blob/master/packages/plugins/content/image/src/createPlugin.tsx#L20

the `Component` of a custom controls currently receives the same props as the `Renderer`: https://github.com/react-page/react-page/blob/master/packages/editor/src/core/types/plugins.ts#L85

which is declared here: https://github.com/react-page/react-page/blob/master/packages/editor/src/core/types/plugins.ts#L9

basically it receives `data` (the value) and `onChange`, like a typical controlled input. Show the values of `data` in your custom controls and if you want to save changes, call `onChange` with the new data.

depending on your usecase, its easier to use the `autoform`-control type, you can still customize the layout with https://github.com/react-page/react-page/blob/master/packages/editor/src/core/types/plugins.ts#L110

**Reply by zehawki** [reply]
Hey @macrozone  thank you! I have it all up and running already :-). After some initial hiccups, it was easy.

> depending on your usecase, its easier to use the autoform-control type

Well, what I needed was to be able to create a drop down selector with dynamic values (based on latest info coming from the redux store) - and I could find no method to do this in the autoform control. And then next, I want to make the drop down selector into a graphical selector, so thats its visually more attractive.

**Reply by macrozone** [reply]
did this all the time, no problems with that.

you can customize the components of any field in the schema: https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-field-components-example

your component can then fetch options from a graphql api, from redux or anything


See also this example here https://github.com/react-page/react-page/blob/master/examples/pages/examples/reactadmin.tsx#L209
where it uses react admin and react admin inputs to fetch options of a list (in this case products)

**Reply by zehawki** [reply]
> you can customize the components of any field in the schema: https://react-page.github.io/docs/#/custom-cell-plugins?id=custom-field-components-example

Yup, I've seen that earlier and still need to play with it to understand properly how it works. For instance, I don't understand how to return an object. In the examples Ive seen (as pointed above), the custom component returns a string. If I want to return an object, should I just say `type: 'object'` and will it automatically work?

```
selectComponent: {
  type: 'object',
  uniforms: {
    // you can pass additional props to uniforms, e.g. the component to use to render the field
    component: MyCustomComponent,
  },
},
```

**Reply by macrozone** [reply]
yes, in the schema, use `type: "object` and then you can define the shape of the object with `properties`.

your custom component will receive the object as `value`, pass a new object on any change with `onChange`
